arbor/morph/segment_tree.cpp

```diff
@@ -10,72 +10,103 @@
+                                                              const node_t& start,
+                                                              id_p predicate,
+                                                              const segment_tree& init={}) {
     auto children_of = tree_to_children(tree);
     auto& segments = tree.segments();
     segment_tree result = init;
+    std::vector<bool> copied(segments.size(), false);
     auto todo = std::vector<node_t>{start};
     while (!todo.empty()) {
         auto node = todo.back();
         todo.pop_back();
-        if (!predicate(node)) continue;
+        if (!predicate(node.id)) continue;
+        copied[node.id] = true;
         const auto& segment = segments[node.id];
         auto current = result.append(node.parent, segment.prox, segment.dist, segment.tag);
         for (auto child: children_of[node.id]) {
             todo.push_back({current, child});
         }
     }
-    return result;
+    return {result, copied};
+}
+
+// Filter a segment tree as a graph
+// - tree to be (partially) filtered
+// - predicate: if returning false for a given node, we exclude that node from the
+//              returned tree. Can be used to filter trees by segment id.
+std::pair<segment_tree, std::vector<bool>> copy_fulltree_if(const segment_tree& tree,
+                                                            id_p predicate) {
+    auto& segments = tree.segments();
+    auto& parents = tree.parents();
+
+    segment_tree result;
+    std::vector<bool> copied(segments.size(), false);
+    std::vector<msize_t> new_ids(segments.size(), mnpos);
+
+    for (msize_t id: make_span(0, segments.size())) {
+        if (!predicate(id)) continue;
+        copied[id] = true;
+        const auto& segment = segments[id];
+        auto parent = parents[id];
+        new_ids[id] = result.append(parent != mnpos ? new_ids[parent] : mnpos,
+                                    segment.prox, segment.dist, segment.tag);
+    }
+    return {result, copied};
 }
 
 std::pair<segment_tree, segment_tree>
 split_at(const segment_tree& tree, msize_t at) {
     if (at >= tree.size() || at == mnpos) throw invalid_segment_parent(at, tree.size());
     // span the sub-tree starting at the splitting node
-    segment_tree post = copy_if(tree, {mnpos, at}, yes);
-    // copy the original tree, but skip all nodes in the `post` tree
-    segment_tree pre = copy_if(tree,
-                               {mnpos, 0},
-                               [=](auto& node) { return node.id != at; });
+    const auto& post_copied = copy_subtree_if(tree, {mnpos, at}, yes);
+    auto post = post_copied.first;
+    auto copied = post_copied.second;
+
+    // copy the original segment_tree (as a graph), skipping all nodes in the `post` subtree
+    segment_tree pre = copy_fulltree_if(tree,
+                                        [=](msize_t id) { return !copied[id]; }).first;
```

Comment:
Consuming `copied` by value?
```suggestion
                                        [&](msize_t id) { return !copied[id]; }).first;
```
Better?

Reply:
Yes, integrated in dfaba6b.